snafu/benchmarks/_load_benchmarks.py

```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Holds function for automagically importing all benchmark modules in ``snafu.benchmarks``
+
+Assumes that each module under ``snafu.benchmarks`` contains a class that subclasses ``Benchmark``.
+"""
+import sys
+import os
+import pkgutil
+import importlib
+import traceback
+import logging
+from dataclasses import dataclass
+from typing import Dict, Tuple, Type, List, Union
+from types import TracebackType
+
+
+_EXC_INFO_TYPE = Union[Tuple[Type[BaseException], BaseException, TracebackType], Tuple[None, None, None]]
+
+
+@dataclass
+class DetectedBenchmarks:
+    imported: List[str]
+    failed: List[str]
+    errors: Dict[str, _EXC_INFO_TYPE]
+
+    def log(self, logger: logging.Logger, level: int = logging.DEBUG, show_tb: bool = False) -> None:
+        logger.log(
+            level,
+            f"Successfully imported {len(self.imported)} benchmark modules: {', '.join(self.imported)}",
+        )
+        logger.log(
+            level, f"Failed to import {len(self.failed)} benchmark modules: {', '.join(self.failed)}",
+        )
+        if show_tb and len(self.errors) > 0:
+            logger.log(level, f"Got the following errors:")
+            for benchmark, exc_info in self.errors.items():
+                tb = "".join(traceback.format_exception(*exc_info))
+                logger.log(level, f"Benchmark module {benchmark} failed to import:\n{tb}")
+
+
+def load_benchmarks() -> DetectedBenchmarks:
+    """
+    Autodetect modules in same directory as source file (``__file__``) and automatically import them.
```

Comment:
This replaces the static method of benchmark discovery that we currently have in ``utils/wrapper_factory.py``, which keeps a mapping of benchmark names to wrapper objects. My thinking was that this would reduce maintenance as we don't have to keep this file up-to-date as benchmarks are added and removed. All a user has to do is place their module under ``benchmarks``, no other code has to be modified. Additionally, this dynamic import system means that we don't have to ship snafu with dependencies for each benchmark. The current wrapper factory requires importing each benchmark, which means each benchmark's python dependencies have to be present, otherwise an import error is raised. This new method will catch these import errors, log them, and only expose modules that were successfully imported.